File: app/src/main/java/io/ipoli/android/player/data/Inventory.java
package io.ipoli.android.player.data;

import com.fasterxml.jackson.annotation.JsonIgnore;

import org.threeten.bp.LocalDate;

import java.util.HashMap;
import java.util.Map;

import io.ipoli.android.store.PowerUp;

import static io.ipoli.android.app.utils.DateUtils.toMillis;

/**
 * Created by Polina Zhelyazkova <polina@ipoli.io>
 * on 5/25/17.
 */

public class Inventory {
    private Map<Integer, Long> powerUps;
    private Map<Integer, Long> pets;
    private Map<Integer, Long> avatars;

    public Inventory() {
        // intentional
    }

    @JsonIgnore
    public void addPowerUp(PowerUp powerUp, LocalDate expirationDate) {
        getPowerUps().put(powerUp.code, toMillis(expirationDate));
    }

    @JsonIgnore
    public void addPet(PetAvatar petAvatar, LocalDate localDate) {
        getPets().put(petAvatar.code, toMillis(localDate));
    }

    @JsonIgnore
    public void addAvatar(Avatar avatar, LocalDate localDate) {
        getAvatars().put(avatar.code, toMillis(localDate));
    }

    public Map<Integer, Long> getPowerUps() {
        if (powerUps == null) {
            powerUps = new HashMap<>();
        }
        return powerUps;
    }

    public void setPowerUps(Map<Integer, Long> powerUps) {
        this.powerUps = powerUps;
    }

    public Map<Integer, Long> getPets() {
        if (pets == null) {
            pets = new HashMap<>();
        }
        return pets;
    }

    public void setPets(Map<Integer, Long> pets) {
        this.pets = pets;
    }

    public Map<Integer, Long> getAvatars() {
        if (avatars == null) {
            avatars = new HashMap<>();
        }
        return avatars;
    }

    public void setAvatars(Map<Integer, Long> avatars) {
        this.avatars = avatars;
    }

    @JsonIgnore
    public void enableAllPowerUps(LocalDate expirationDate) {
        for (PowerUp powerUp : PowerUp.values()) {
            addPowerUp(powerUp, expirationDate);
        }
    }
}

File: app/src/main/java/io/ipoli/android/player/data/Inventory.kt
package io.ipoli.android.player.data

import io.ipoli.android.common.datetime.DateUtils
import io.ipoli.android.store.avatars.data.Avatar
import io.realm.RealmList
import io.realm.RealmObject
import org.threeten.bp.LocalDate


/**
 * Created by Polina Zhelyazkova <polina@ipoli.io>
 * on 8/22/17.
 */
open class Inventory() : RealmObject() {

    var powerUps: RealmList<InventoryItem> = RealmList()
    var pets: RealmList<InventoryItem> = RealmList()
    var avatars: RealmList<InventoryItem> = RealmList()

    //    fun addPowerUp(powerUp: PowerUp, expirationDate: LocalDate) {
//        getPowerUps().put(powerUp.code, toMillis(expirationDate))
//    }
//
//    fun addPet(petAvatar: PetAvatar, localDate: LocalDate) {
//        getPets().put(petAvatar.code, toMillis(localDate))
//    }

    fun addAvatar(avatar: Avatar, localDate: LocalDate) {
        avatars.add(InventoryItem(avatar.code, DateUtils.toMillis(localDate)))
    }

    fun addAvatar(code: Int, localDate: LocalDate) {
        avatars.add(InventoryItem(code, DateUtils.toMillis(localDate)))
    }

    fun hasAvatar(code: Int): Boolean {
        return avatars.find { inventoryItem -> inventoryItem.code == code } != null
    }
//    fun enableAllPowerUps(expirationDate: LocalDate) {
//        for (powerUp in PowerUp.values()) {
//            addPowerUp(powerUp, expirationDate)
//        }
//    }
}

open class InventoryItem(var code: Int = -1, var date: Long = -1) : RealmObject() {
}